subsys/bluetooth/host/hci_core.c

```diff
@@ -450,13 +453,24 @@ static void rpa_timeout(struct k_work *work)
 	if (atomic_test_bit(bt_dev.flags, BT_DEV_ADVERTISING)) {
 		/* make sure new address is used */
 		set_advertise_enable(false);
-		le_set_private_addr(bt_dev.adv_id);
+		err_adv = le_set_private_addr(bt_dev.adv_id);
 		set_advertise_enable(true);
 	}
 
 	if (atomic_test_bit(bt_dev.flags, BT_DEV_ACTIVE_SCAN)) {
 		/* TODO do we need to toggle scan? */
-		le_set_private_addr(BT_ID_DEFAULT);
+		err_scan = le_set_private_addr(BT_ID_DEFAULT);
+	}
+
+	/* If both advertising and scanning is active, le_set_private_addr
+	 * will fail. In this case, set back RPA_VALID so that if either of
+	 * advertising or scanning was restarted by application then
+	 * le_set_private_addr in the public API call path will not retry
+	 * set_random_address. This is needed so as to be able stop and restart
+	 * either role by applicaion after rpa_timeout.
+	 */
+	if (rpa_valid && (err_adv || err_scan)) {
```

Comment:
Do you need the `rpa_valid` condition here? If this function was invoked with `BT_DEV_RPA_VALID` set to 0, then both adv and scan will still try to set the private address, so why not set it in that case?

Reply:
My point there is that I don't think there is a case where you actually care about whether `BT_DEV_RPA_VALID` was 0 or 1. If the RPA timeout fires and there's roles running, you need to update the RPA, and if that fails because there's another role running then you need to set `BT_DEV_RPA_VALID` to 1. This would save an auto variable and a comparison

Reply:
@jhedberg do you agree with the above?

Reply:
@carlescufi yes, I do